src/vs/workbench/contrib/positronIPyWidgets/browser/positronIPyWidgetsService.ts

```diff
@@ -199,6 +218,104 @@ export class PositronIPyWidgetsService extends Disposable implements IPositronIP
+	 * output widget. See https://ipywidgets.readthedocs.io/en/latest/examples/Output%20Widget.html
+	 * for more.
+	 *
+	 * @param sessionId The runtime session ID.
+	 * @param parentId The parent message ID.
+	 */
+	willHandleMessage(sessionId: string, parentId: string): boolean {
+		return Array.from(this._outputManagersBySessionId.values())
+			// Get all output handlers for the session.
+			.filter(instance => instance.sessionId === sessionId)
+			// Check if any will handle the message.
+			.some(handler => handler.willHandleMessage(parentId));
+	}
+}
+
+/**
+ * Manages IPyWidgets output clients for a given language runtime session.
+ */
+class IPyWidgetsOutputClientManager extends Disposable {
+	/** The session ID. */
+	public readonly sessionId: string;
+
+	/** Map of IPyWidget output clients, keyed by client ID. */
+	private readonly _clients = this._register(new DisposableMap<string, IPyWidgetsOutputClientInstance>());
+
+	constructor(session: ILanguageRuntimeSession) {
+		super();
+
+		this.sessionId = session.sessionId;
+
+		// Create output clients for existing runtime clients.
+		if (session.getRuntimeState() !== RuntimeState.Uninitialized) {
+			session.listClients(RuntimeClientType.IPyWidget).then((clients) => {
+				for (const client of clients) {
+					this.createClient(client);
+				}
+			});
+		}
+
+		// Create output clients for new runtime clients.
+		this._register(session.onDidCreateClientInstance(({ client }) => {
+			// Only handle IPyWidget clients.
+			if (client.getClientType() !== RuntimeClientType.IPyWidget) {
+				return;
+			}
+
+			// Create and register the client.
+			this.createClient(client);
+		}));
+	}
+
+	private createClient(client: IRuntimeClientInstance<any, any>) {
+		const clientId = client.getClientId();
+		const outputClient = new IPyWidgetsOutputClientInstance(client);
+		this._clients.set(clientId, outputClient);
+
+		// Delete and dispose the output client when it is closed.
+		this._register(Event.fromObservable(client.clientState)(state => {
+			if (state === RuntimeClientState.Closed) {
+				this._clients.deleteAndDispose(clientId);
+			}
+		}));
+	}
+
+	/** Whether any of this session's clients will handle messages to the given parent ID. */
+	willHandleMessage(parentId: string) {
+		return Array.from(this._clients.values())
+			.some(client => client.willHandleMessage(parentId));
+	}
+}
+
+/**
+ * An IPyWidgets Output widget client instance.
+ */
+class IPyWidgetsOutputClientInstance extends Disposable {
+	/** The parent message ID that this widget client will handle, if any. */
+	private _willHandleMessagesForParentId?: string;
+
+	constructor(client: IRuntimeClientInstance<any, any>) {
+		super();
+
+		this._register(client.onDidReceiveData(({ data }) => {
+			// If this is an output widget beginning to handle messages, remember the parent ID.
+			if (data?.method === 'update' && typeof data?.state?.msg_id === 'string') {
```

Comment:
Alternatively, you could also use a predicate function here for the if to clean it up:

```ts
function isUpdateMsgData(data: unknown): data is {
    method: 'update';
    state: { msg_id: string } ;
} {
    return (
        typeof data === 'object' &&
        data !== null &&
        'method' in data &&
        data.method === 'update' &&
        'state' in data &&
        typeof data.state === 'object' &&
        data.state !== null &&
        'msg_id' in data.state &&
        typeof data.state.msg_id === 'string'
    );
}```